jni/src/Foreign/JNI/Types.hs

```diff
@@ -133,6 +136,16 @@ instance IsReferenceType ('Iface sym)
 instance IsReferenceType ('Array ty)
 instance IsReferenceType ty => IsReferenceType ('Generic ty tys)
 
+-- | Produces evidence for IsReferenceType from a `Sing ty`.
+fromReferenceType :: Sing (ty :: JType) -> (IsReferenceType ty => a) -> Maybe a
```

Comment:
I find this name confusing. What would we call the converse function? `toReferenceType :: ... -> Sing ty`? If anything I'd call *this* function `singToReferenceType`. Regarding type signature, I think the following would be more intuitive:

```
singToReferenceType :: Sing ty -> Maybe (Dict (IsReferenceType ty))
```

The extra dependency on `constraints` this implies is pretty benign.